File: src/sc_neurocore/accel/julia/network/population.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia acceleration for network/population

module PopulationAccel

using Statistics, LinearAlgebra

mutable struct PopulationState
    neurons::Float64
    n::Float64
    model_name::Float64
    label::Float64
    _model_cls::Float64
    _voltages::Float64
end

function PopulationState()
    PopulationState(0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
end

function _sync_voltages(s::PopulationState)
    for i, neuron in enumerate(s.neurons)
        s._voltages[i] = getattr(neuron, "v", 0.0)
end

function step_all(s::PopulationState, currents, spike_gating)
    spikes = zeros(s.n, dtype=np.int8)
    if spike_gating
        for i, neuron in enumerate(s.neurons)
            v = getattr(neuron, "v", 0.0)
            v_thresh = getattr(neuron, "v_threshold", 1.0)
            v_rest = getattr(neuron, "v_rest", 0.0)
            # Skip if no input AND voltage within 1% of rest
            if currents[i] == 0.0 && abs(v - v_rest) < 0.01 * abs(v_thresh - v_rest)
                continue
            raw = neuron.step(float(currents[i]))
            spikes[i] = min(max(int(raw), 0), 1)
            s._voltages[i] = getattr(neuron, "v", 0.0)
    else
        for i, neuron in enumerate(s.neurons)
            raw = neuron.step(float(currents[i]))
            spikes[i] = min(max(int(raw), 0), 1)
            s._voltages[i] = getattr(neuron, "v", 0.0)
    return spikes
end

function reset_all(s::PopulationState)
    for neuron in s.neurons
        if hasattr(neuron, "reset")
            neuron.reset()
        elseif hasattr(neuron, "reset_state")
            neuron.reset_state()
    s._sync_voltages()
end

function get_states(s::PopulationState)
    if s.n == 0
        return {}
    sample = s.neurons[0]
    if hasattr(sample, "get_state")
        keys = sample.get_state().keys()
    elseif hasattr(sample, "__dataclass_fields__")
        keys = [k for k in sample.__dataclass_fields__ if k ! in ("dt",)]
    else
        keys = ["v"]
    result = {}
    for k in keys
        result[k] = collect([getattr(n, k, 0.0) for n in s.neurons])
    return result
end

function set_voltages(s::PopulationState, voltages)
    for i, neuron in enumerate(s.neurons)
        if hasattr(neuron, "v")
            neuron.v = float(voltages[i])
    s._voltages[:] = voltages[: s.n]
end

function voltages(s::PopulationState)
    return s._voltages
end

end # module PopulationAccel

File: src/sc_neurocore/accel/rust/safety/population.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for population

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct Population {
    pub neurons: f64,
    pub n: f64,
    pub model_name: f64,
    pub label: f64,
    pub _model_cls: f64,
    pub _voltages: f64,
}

impl Population {
    pub fn new() -> Self {
        Self {
            neurons: 0.0_f64,
            n: 0.0_f64,
            model_name: 0.0_f64,
            label: 0.0_f64,
            _model_cls: 0.0_f64,
            _voltages: 0.0_f64,
        }
    }

    pub fn _sync_voltages(&self, ) -> f64 {
        // for i, neuron in enumerate(self.neurons):
        // self._voltages[i] = getattr(neuron, "v", 0.0)
        0.0
    }

    pub fn step_all(&self, currents: f64, spike_gating: f64) -> f64 {
        // spikes = np.zeros(self.n, dtype=np.int8)
        // if spike_gating:
        // for i, neuron in enumerate(self.neurons):
        // v = getattr(neuron, "v", 0.0)
        // v_thresh = getattr(neuron, "v_threshold", 1.0)
        // v_rest = getattr(neuron, "v_rest", 0.0)
        // # Skip if no input AND voltage within 1% of rest
        // if currents[i] == 0.0 && abs(v - v_rest) < 0.01 * abs(v_thresh - v_res
        // continue
        // raw = neuron.step(float(currents[i]))
        // spikes[i] = min(max(int(raw), 0), 1)
        // self._voltages[i] = getattr(neuron, "v", 0.0)
        // else:
        // for i, neuron in enumerate(self.neurons):
        // raw = neuron.step(float(currents[i]))
        0.0
    }

    pub fn reset_all(&self, ) -> f64 {
        // for neuron in self.neurons:
        // if hasattr(neuron, "reset"):
        // neuron.reset()
        // elif hasattr(neuron, "reset_state"):
        // neuron.reset_state()
        // self._sync_voltages()
        0.0
    }

    pub fn get_states(&self, ) -> f64 {
        // if self.n == 0:
        // return {}
        // sample = self.neurons[0]
        // if hasattr(sample, "get_state"):
        // keys = sample.get_state().keys()
        // elif hasattr(sample, "__dataclass_fields__"):
        // keys = [k for k in sample.__dataclass_fields__ if k not in ("dt",)]
        // else:
        // keys = ["v"]
        // result = {}
        // for k in keys:
        // result[k] = np.array([getattr(n, k, 0.0) for n in self.neurons])
        // return result
        0.0
    }

    pub fn set_voltages(&self, voltages: f64) -> f64 {
        // for i, neuron in enumerate(self.neurons):
        // if hasattr(neuron, "v"):
        // neuron.v = float(voltages[i])
        // self._voltages[:] = voltages[: self.n]
        0.0
    }

    pub fn voltages(&self, ) -> f64 {
        // return self._voltages
        0.0
    }

}

pub fn validate_population(state: &Population) -> bool {
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_population_new() {
        let state = Population::new();
        assert!(validate_population(&state));
    }

}
